draft-vvv-webtransport-quic.md

```diff
@@ -154,62 +154,70 @@ datagram extension [QUIC-DATAGRAM].
-sent a client indication.  From the server perspective, the session is
-established after the client indication has been successfully processed.
+sent its headers.  From the server perspective, the session is
+established after the client headers have been successfully processed.
 
 ## Identifying as QuicTransport
 
 In order to identify itself as a WebTransport application, QuicTransport relies
 on TLS Application-Layer Protocol Negotiation {{!RFC7301}}.  The user agent MUST
-request the ALPN value of "wq-vvv-01" and it MUST close the connection unless
+request the ALPN value of "wq-vvv-03" and it MUST close the connection unless
 the server confirms that ALPN value.
 
-## Client Indication
+## QuicTransport Handshake
 
 In order to verify that the client's origin is allowed to connect to the server
 in question, the user agent has to communicate the origin to the server.  This
-is accomplished by sending a special message, called client indication, on
-stream 2, which is the first client-initiated unidirectional stream.
+is accomplished by performing a handshake on the first client-initiated
+bidirectional stream, stream 0.  The handshake is performed as follows:
 
-The client indication is a sequence of key-value pairs that are formatted in the
-following way:
+1. The client sends its headers on stream 0.
+1. The client sends FIN on stream 0.  This indicates that all headers have been
+   sent, and the server can now process them.
+1. The server processes the headers and makes access control decision based on
+   the origin.  Before this happens, the server MUST NOT process the
+   application data receiver from the client, or send its own header block.
+1. The server sends its header block on stream 0, followed by a FIN indicating
+   that the handshake has been successfully completed.
+
+The QuicTransport header stream payload SHALL be a header block.  A header
+block is a sequence of key-value pairs that are formatted in the following way:
 
 ~~~
- 0                   1                   2                   3
- 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|           Key (16)            |          Length (16)          |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                           Value (*)                         ...
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+QuicTransport Header {
+  Header Name Length (16),
+  Header Name (..),
+  Header Value Length (16),
+  Header Value (..),
+}
 ~~~
-{: #fig-frame title="Client indication format" :}
+{: #fig-frame title="QuicTransport Header Format" :}
 
 The pair includes the following fields:
 
-Key:
-
-: Indicates the field that is being expressed.
-
-Length:
+Name:
 
-: Indicates the length of the value (the length of the key and the length
-  itself are not included).
+: Indicates the name of the field that is being expressed.  The name of the
```

Comment:
Why is this more restrictive than HTTP headers? Couldn't we use the same rules as HTTP/2 headers for simplicity?

Reply:
This is actually based on [the most restrictive interpretation of HTTP header name restrictions](https://httpwg.org/http-core/draft-ietf-httpbis-semantics-latest.html#rfc.section.5.3.p.6).

Reply:
I think sticking to HTTP semantics would be best. Adding a reference to that link SGTM as a way to suggest folks use this subset.